tests/integration/ha_tests/test_storage.py

```diff
@@ -0,0 +1,343 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+import os
+import time
+
+import pytest
+from juju import tag
+from pytest_operator.plugin import OpsTest
+
+from ..helpers import check_or_scale_app, get_app_name
+from .helpers import (
+    APP_NAME,
+    count_writes,
+    reused_storage,
+    stop_continous_writes,
+    storage_id,
+    storage_type,
+)
+
+OTHER_APP_NAME = "mongodb-new"
+TIMEOUT = 1000
+
+
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.skipif(
+    os.environ.get("PYTEST_SKIP_DEPLOY", False),
+    reason="skipping deploy, model expected to be provided.",
+)
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, database_charm) -> None:
+    """Build and deploy one unit of MongoDB."""
+    # it is possible for users to provide their own cluster for HA testing. Hence check if there
+    # is a pre-existing cluster.
+    required_units = 3
+    user_app_name = await get_app_name(ops_test)
+    if user_app_name:
+        await check_or_scale_app(ops_test, user_app_name, required_units)
+        return
+
+    storage = {"mongodb": {"pool": "lxd", "size": 2048}}
+
+    await ops_test.model.deploy(
+        database_charm, application_name=APP_NAME, num_units=required_units, storage=storage
+    )
+    await ops_test.model.wait_for_idle()
+
+
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_storage_re_use(ops_test, continuous_writes):
+    """Verifies that database units with attached storage correctly repurpose storage.
+
+    It is not enough to verify that Juju attaches the storage. Hence test checks that the mongod
+    properly uses the storage that was provided. (ie. doesn't just re-sync everything from
+    primary, but instead computes a diff between current storage and primary storage.)
+    """
+    app_name = await get_app_name(ops_test)
+    if storage_type(ops_test, app_name) == "rootfs":
+        pytest.skip(
+            "reuse of storage can only be used on deployments with persistent storage not on rootfs deployments"
+        )
+
+    # removing the only replica can be disastrous
+    if len(ops_test.model.applications[app_name].units) < 2:
+        await ops_test.model.applications[app_name].add_unit(count=1)
+        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=TIMEOUT)
```

Comment:
I think this case be removed since the previous test scales the provided deployment?